apps/bfd-db-migrator/src/test/java/gov/cms/bfd/migrator/util/DatabaseTestUtils.java

```diff
@@ -0,0 +1,427 @@
+    this.unpooledDataSource = initUnpooledDataSource();
+  }
+
+  /** @return the singleton {@link DatabaseTestUtils} instance to use everywhere */
+  public static synchronized DatabaseTestUtils get() {
+    /*
+     * Why are we using a singleton and caching all of these fields? Because creating some of the
+     * fields here is expensive, so we don't want to have to re-create it for every test.
+     */
+
+    if (SINGLETON == null) {
+      SINGLETON = new DatabaseTestUtils();
+    }
+
+    return SINGLETON;
+  }
+
+  /**
+   * @return the {@link DataSource} for the database to test against (as specified by the <code>
+   *     its.db.*</code> system properties, see {@link #initUnpooledDataSource() for details})
+   */
+  private static DataSource initUnpooledDataSource() {
+    /*
+     * This is pretty hacky, but when this class is being used as part of the BFD Server tests, we
+     * have to check for the DB connection properties that the BFD Server may have written out when
+     * it was launched for the ITs run. If we DON'T use those properties, we'll end up connected to
+     * a different database than the one that the application server instance being tested is using,
+     * which is definitely not going to do what we wanted.
+     */
+    Optional<Properties> bfdServerTestDatabaseProperties = readTestDatabaseProperties();
+
+    String url, username, password;
+    if (bfdServerTestDatabaseProperties.isPresent()) {
+      url = bfdServerTestDatabaseProperties.get().getProperty("bfdServer.db.url");
+      username = bfdServerTestDatabaseProperties.get().getProperty("bfdServer.db.username");
+      password = bfdServerTestDatabaseProperties.get().getProperty("bfdServer.db.password");
+    } else {
+      /*
+       * Build default DB connection properties that use HSQL, just as they're configured in the
+       * parent POM.
+       */
+      String urlDefault = String.format("%shsqldb:mem", JDBC_URL_PREFIX_BLUEBUTTON_TEST);
+      String usernameDefault = null;
+      String passwordDefault = null;
+
+      // Build the actual DB connection properties to use.
+      url = System.getProperty("its.db.url", urlDefault);
+
+      username = System.getProperty("its.db.username", usernameDefault);
+      if (username != null && username.trim().isEmpty()) username = usernameDefault;
+      password = System.getProperty("its.db.password", passwordDefault);
+      if (password != null && password.trim().isEmpty()) password = passwordDefault;
+    }
+
+    return initUnpooledDataSource(url, username, password);
+  }
+
+  /**
+   * @return the {@link Properties} file that contains the test DB connection properties (as created
+   *     by <code>gov.cms.bfd.server.war.SpringConfiguration#findTestDatabaseProperties()</code>, or
+   *     {@link Optional#empty()} if it's not present (indicating that just a regular DB connection
+   *     is being used)
+   */
+  private static Optional<Properties> readTestDatabaseProperties() {
```

Comment:
This is a general comment for the properties; ideally I'd also like to see some sort of flyway config properties file. For example, one thing I constantly bang my head is the inability to mix transactional and non-transaction sql in a given migration. Fly allows you to enable/disable this via config, but since we don't support an actual flyway config we end up with sometimes multiple flyway scripts to perform one fast easy functional change.

Also, not sure, but has anyone looked at flyway, the app. It supports significant functionality via a combination of cmd-line args and config file(s).

Reply:
> I'd also like to see some sort of flyway config properties file.
I think that could be great!

> Also, not sure, but has anyone looked at flyway, the app. It supports significant functionality via a combination of cmd-line args and config file(s).

We had considered the `flyway` cli utility, and that idea may still have merit. Obviously, there's more than one way to do this kind of thing. For me personally, I started thinking through how we'd integrate an external call to `flyway` with our current testing strategy and the potential for continued hibernate validations here, and that gave me _some_ pause... what do you think the alternative would look like?

Reply:
I'll look into supporting a flyway config; I'm not sure what needs to be done to allow that, but I can tinker with it. (This might be worth a side story though, it wasn't part of this initial scope)

Reply:
I opened [BFD-1625](https://jira.cms.gov/browse/BFD-1625) to add configuration file support to the app; let's chat soon and figure out how I could best implement this for what you're thinking of doing with it.